File: src/main/java/no/novari/flyt/authorization/user/controller/MeController.java
package no.novari.flyt.authorization.user.controller;

import no.novari.flyt.authorization.client.sourceapplications.model.SourceApplication;
import no.novari.flyt.authorization.user.UserService;
import no.novari.flyt.authorization.user.controller.utils.TokenParsingUtils;
import no.novari.flyt.authorization.user.model.RestrictedPageAuthorization;
import no.novari.flyt.authorization.user.model.User;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static no.novari.flyt.resourceserver.UrlPaths.INTERNAL_API;

@RestController
@RequestMapping(INTERNAL_API + "/authorization/me")
public class MeController {

    private final TokenParsingUtils tokenParsingUtils;
    private final UserService userService;
    private final List<SourceApplication> sourceApplications;

    public MeController(
            TokenParsingUtils tokenParsingUtils,
            UserService userService,
            List<SourceApplication> sourceApplications
    ) {
        this.tokenParsingUtils = tokenParsingUtils;
        this.userService = userService;
        this.sourceApplications = sourceApplications;
    }

    @GetMapping("is-authorized")
    public ResponseEntity<?> checkAuthorization(
            @AuthenticationPrincipal Authentication authentication
    ) {
        JwtAuthenticationToken jwtAuthToken = (JwtAuthenticationToken) authentication;
        Optional<User> userOptional = getUserFromUserAuthorizationComponent(jwtAuthToken);

        if (userOptional.isEmpty() && tokenParsingUtils.hasPermittedRole(jwtAuthToken)) {
            createUserFromToken(authentication);
        }
        return ResponseEntity.ok("User authorized");
    }

    @GetMapping("restricted-page-authorization")
    public ResponseEntity<RestrictedPageAuthorization> getRestrictedPageAuthorization(
            @AuthenticationPrincipal Authentication authentication
    ) {
        boolean isAdmin = tokenParsingUtils.isAdmin(authentication);
        RestrictedPageAuthorization authorization = RestrictedPageAuthorization
                .builder()
                .userPermissionPage(isAdmin)
                .build();
        return ResponseEntity.ok(authorization);
    }

    @GetMapping
    public ResponseEntity<User> get(
            @AuthenticationPrincipal Authentication authentication
    ) {
        JwtAuthenticationToken jwtAuthToken = (JwtAuthenticationToken) authentication;
        Optional<User> userOptional = getUserFromUserAuthorizationComponent(jwtAuthToken);
        if (userOptional.isPresent()) {
            return ResponseEntity.ok(userOptional.get());
        } else {
            User newUser = createUserFromToken(authentication);
            return ResponseEntity.ok(newUser);
        }
    }

    private Optional<User> getUserFromUserAuthorizationComponent(JwtAuthenticationToken token) {
        return userService.find(UUID.fromString(tokenParsingUtils.getObjectIdentifierFromToken(token)));
    }

    private User createUserFromToken(Authentication authentication) {
        JwtAuthenticationToken jwtAuthToken = (JwtAuthenticationToken) authentication;
        User.UserBuilder userBuilder = tokenParsingUtils.getUserFromToken(jwtAuthToken).toBuilder();
        boolean isAdmin = tokenParsingUtils.isAdmin(authentication);
        if (isAdmin) {
            userBuilder.sourceApplicationIds(allSourceApplicationIds());
        }
        User newUser = userBuilder.build();
        userService.save(newUser);
        return newUser;
    }

    private List<Long> allSourceApplicationIds() {
        return sourceApplications.stream()
                .map(SourceApplication::getId)
                .sorted()
                .toList();
    }

}

File: src/main/kotlin/no/novari/flyt/authorization/user/controller/MeController.kt
package no.novari.flyt.authorization.user.controller

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import no.novari.flyt.authorization.client.sourceapplications.model.SourceApplication
import no.novari.flyt.authorization.user.UserService
import no.novari.flyt.authorization.user.controller.utils.TokenParsingUtils
import no.novari.flyt.authorization.user.model.RestrictedPageAuthorization
import no.novari.flyt.authorization.user.model.User
import no.novari.flyt.webresourceserver.UrlPaths.INTERNAL_API
import org.springframework.http.HttpStatus
import org.springframework.security.core.Authentication
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException

@RestController
@RequestMapping("$INTERNAL_API/authorization/me")
@Tag(name = "Current user", description = "Authorization and permissions for the authenticated user.")
class MeController(
    private val tokenParsingUtils: TokenParsingUtils,
    private val userService: UserService,
    private val sourceApplications: List<SourceApplication>,
) {
    @GetMapping("is-authorized")
    @Operation(summary = "Verify that the current user is authorized")
    fun checkAuthorization(authentication: Authentication?): String {
        val jwtAuthToken = requireJwtAuthenticationToken(authentication)
        if (tokenParsingUtils.hasPermittedRole(jwtAuthToken)) {
            userService.findOrCreate(buildUserFromToken(jwtAuthToken, authentication))
        }

        return "User authorized"
    }

    @GetMapping("restricted-page-authorization")
    @Operation(summary = "Get restricted-page authorization for the current user")
    fun getRestrictedPageAuthorization(authentication: Authentication?): RestrictedPageAuthorization {
        return RestrictedPageAuthorization(
            userPermissionPage = tokenParsingUtils.isAdmin(authentication),
        )
    }

    @GetMapping
    @Operation(summary = "Get or provision the current user")
    fun get(authentication: Authentication?): User {
        val jwtAuthToken = requireJwtAuthenticationToken(authentication)
        return userService.findOrCreate(buildUserFromToken(jwtAuthToken, authentication))
    }

    private fun buildUserFromToken(
        jwtAuthToken: JwtAuthenticationToken,
        authentication: Authentication?,
    ): User {
        val user = tokenParsingUtils.getUserFromToken(jwtAuthToken)
        return if (tokenParsingUtils.isAdmin(authentication)) {
            user.copy(sourceApplicationIds = allSourceApplicationIds())
        } else {
            user
        }
    }

    private fun allSourceApplicationIds(): List<Long> {
        return sourceApplications.map(SourceApplication::id).sorted()
    }

    private fun requireJwtAuthenticationToken(authentication: Authentication?): JwtAuthenticationToken {
        return authentication as? JwtAuthenticationToken
            ?: throw ResponseStatusException(HttpStatus.UNAUTHORIZED, "Authentication is required")
    }
}
